torchx/schedulers/ray/ray_driver.py

```diff
@@ -0,0 +1,144 @@
+        pg = ray.util.placement_group(bundles, strategy="SPREAD")
+        pgs.append(pg)
+
+        _logger.info("Waiting for placement group to start.")
+        ready = pg.wait(timeout_seconds=100)
+
+        if ready:
+            _logger.info("Placement group has started.")
+            _logger.info("Starting remote function")
+        else:  # pragma: no cover
+            raise TimeoutError(
+                "Placement group creation timed out. Make sure "
+                "your cluster either has enough resources or use "
+                "an autoscaling cluster. Current resources "
+                "available: {}, resources requested by the "
+                "placement group: {}".format(ray.available_resources(), pg.bundle_specs)
+            )
+    return pgs
+
+
+def create_command_actors(
+    actors: List[RayActor], pgs: List[PlacementGroup]
+) -> List[CommandActor]:
+    command_actors: List[CommandActor] = []
+    address, port = get_address_and_port()
+    for i in range(len(actors)):
+        world_size = actors[i].num_replicas
+
+        for rank in range(world_size):
+
+            # Environment variables for distributed training
+            rank_env = {
+                "WORLD_SIZE": str(world_size),
+                "MASTER_PORT": str(port),
+                "MASTER_ADDR": address,
+                "RANK": str(rank),
+            }
+
+            actor_and_rank_env = {**actors[i].env, **rank_env}
+
+            command_actors.append(
+                CommandActor.options(  # pyre-ignore[16]
+                    placement_group=pgs[i],
+                    num_cpus=actors[i].num_cpus,
+                    num_gpus=actors[i].num_gpus,
+                ).remote(actors[i].command, actor_and_rank_env)
+            )
+    return command_actors
+
+
+if __name__ == "__main__":  # pragma: no cover
+    _logger.info("Reading actor.json")
+    actors: List[RayActor] = load_actor_json("actors.json")
+
+    _logger.info("Creating Ray placement groups")
+    ray.init(address="auto", namespace="torchx-ray")
+    pgs: List[PlacementGroup] = create_placement_groups(actors)
+
+    _logger.info("Getting command actors")
+    command_actors: List[CommandActor] = create_command_actors(actors, pgs)
+
+    _logger.info("Running Ray actors")
+    unfinished = [  # pyre-ignore
+        command_actor.run_command.remote()  # pyre-ignore
+        for command_actor in command_actors
+    ]
+
+    # Await return result of remote ray function
+    while len(unfinished) > 0:
+        finished, unfinished = ray.wait(unfinished)
+        # If a failure occurs the ObjectRef will be marked as finished.
+        # Calling ray.get will expose the failure as a RayActorError.
+        for object_ref in finished:
+            try:
+                ray.get(object_ref)
+                _logger.info("Ray remote function promise succesfully returned")
+            except ray.exceptions.RayActorError as exc:
+                _logger.info("Ray Actor Error")
```

Comment:
can we log exception as well? logger.error(..., exc)

Reply:
also we need to aggregate this and print exceptions for all failed actors

Reply:
> also we need to aggregate this and print exceptions for all failed actors

Do you have a favorite way of doing this? I'm thinking of just adding all the exceptions to a list